### PR TITLE
(Closes #1634) Add HoistLocalArraysTrans transformation.

### LINKED ISSUE #1634: Transform automatic arrays to module scope
Automatic arrays cause trouble when attempting to accelerate code on a GPU since, at minimum, they imply an allocation on the GPU every time the routine is called. Even worse, with managed memory, they result in the array being copied from CPU to GPU every time as well. The simplest solution to this problem is to promote automatic arrays to module scope and have them explicitly allocated when they are first accessed. The downside to this is that it increases the code's memory footprint but that could subsequently be improved by using a memory pool for all such work arrays.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
This is ready for a first look now. One for either @sergisiso or @rupertford :-)

**Comment by sergisiso** [comment]
I will do the proper review this afternoon but I tried the branch with NEMO and a couple of errors pop up:

1) This line inside apply fails:
`code = f"allocated({automatic_arrays[0].name})"`
when there are not automatic arrays in the routine. It should return the apply successfully before this or produce a TransformationError during the validation (I prefer the former), but not fail during the apply.

2) It produced invalid code for functions (at least for the nvidia compiler) as it tries to hoist the result value.
```
function alfa_charn(pwnd) result(alfa_charn_1)
    if (.NOT.ALLOCATED(alfa_charn_1)) then
        ALLOCATE(alfa_charn_1(1 : jpi, 1 : jpj))
    end if
```

File: src/psyclone/psyir/frontend/fortran.py [hunk 1]
```diff
@@ -121,6 +121,51 @@ def psyir_from_expression(self, source_code, symbol_table):
                             "into scope.".format(source_code)), err)
         return fake_parent[0].children[0].detach()
 
+    def psyir_from_statement(self, source_code, symbol_table):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fortran.py:124]: Done.

File: src/psyclone/psyir/transformations/hoist_local_arrays_trans.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        container = node.ancestor(Container)
+
+        # Identify all arrays that are local to the target routine and
+        # do not explicitly use dynamic memory allocation.
+        automatic_arrays = []
+        for sym in node.symbol_table.local_datasymbols:
+            if sym.is_array and all(isinstance(dim, ArrayType.ArrayBounds)
+                                    for dim in sym.shape):
+                automatic_arrays.append(sym)
+
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/hoist_local_arrays_trans.py:127]: Thanks :-) Fixed, added test and updated docstring.

File: src/psyclone/psyir/transformations/hoist_local_arrays_trans.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        alloc_arg = ",".join(fwriter(aref) for aref in arefs)
+        body = [freader.psyir_from_statement(f"allocate({alloc_arg})",
+                                             node.symbol_table)]
+        # Insert the conditional allocation at the start of the supplied
+        # routine.
+        node.children.insert(0, IfBlock.create(if_expr, body))
+
+        # Finally, remove the hoisted symbols from the routine scope.
+        for sym in automatic_arrays:
+            # Currently the SymbolTable.remove() method does not support
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/hoist_local_arrays_trans.py:173]: Yes, done.

File: src/psyclone/psyir/transformations/hoist_local_arrays_trans.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            sym.visibility = Symbol.Visibility.PRIVATE
+            # We must allow for the situation where there's a clash with a
+            # symbol already present at container scope.
+            try:
+                container.symbol_table.add(sym)
+            except KeyError:
+                new_name = container.symbol_table.next_available_name(
+                    sym.name, other_table=node.symbol_table)
+                node.symbol_table.rename_symbol(sym, new_name)
+                container.symbol_table.add(sym)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/hoist_local_arrays_trans.py]: Oh dear. You are right. Test added plus a further test that we abort if the parent symbol_table has a clashing tag.

File: src/psyclone/tests/psyir/transformations/hoist_local_arrays_trans_test.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    reader = FortranReader()
+    psyir = reader.psyir_from_source(code)
+    routine = psyir.walk(Routine)[0]
+    hoist_trans = HoistLocalArraysTrans()
+    hoist_trans.apply(routine)
+    code = fortran_writer(psyir).lower()
+    assert ("  real, allocatable, dimension(:,:), private :: a\n"
+            "  real, allocatable, dimension(:,:), private :: a_2\n" in code)
+    assert ("    if (.not.allocated(a_2)) then\n"
+            "      allocate(a_2(1 : nx, 1 : ny))\n"
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/transformations/hoist_local_arrays_trans_test.py:215]: Extended a test. Happily it just works :-)

**Comment by arporter** [comment]
Note to self: I need to add a test for the new method that I've added to SymbolTable (`reverse_tags_dict`).
Probably I should also bite the bullet and test it with NEMOVAR.

**Comment by arporter** [comment]
Testing with NEMOVAR reveals failures where a local array is accessed in a CodeBlock and has been renamed when promoted.

**Comment by arporter** [comment]
The transformation now rejects code if it contains a CodeBlock that references a symbol that is to be hoisted. With this change NEMOVAR processes and validates on CPU (with the transformation being applied to every routine we find).

File: src/psyclone/psyir/transformations/hoist_local_arrays_trans.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+        # CodeBlock (as they may get renamed as part of the transformation).
+        array_names = set(sym.name for sym in auto_arrays)
+        cblocks = node.walk(CodeBlock)
+        for cblock in cblocks:
+            cblock_names = set(cblock.get_symbol_names())
+            if cblock_names.intersection(array_names):
+                cblock_txt = ''.join(str(fpn) for fpn in cblock.get_ast_nodes)
+                raise TransformationError(
+                    f"The supplied routine '{node.name}' contains a local "
+                    f"array '{sym.name}' which is accessed within a CodeBlock "
```
Comment (sergisiso) [inline src/psyclone/psyir/transformations/hoist_local_arrays_trans.py:283]: pylint says: undefined-loop-variable: Using possibly undefined loop variable 'sym'

And btw is really sym.name (the last sym in auto_arrays) what you want here? Or should the `cblock_names.intersection(array_names)` be the list of array names to print?

Comment (sergisiso) [reply src/psyclone/psyir/transformations/hoist_local_arrays_trans.py:283]: Also, do we need to fail? Could this be inside the `_get_local_arrays` and just filter out the ones inside CodeBlocks but continue hoisting the others? Printing a log message would still be useful but we don't do this yet right?

Anyway, its just a suggestion, I leave it up to you to decide what's best.

**Review by sergisiso — CHANGES_REQUESTED** [review]
@arporter It now process NEMO and it improves the OpenMP for GPU performance by 8%. There are just a few more things to iron out but its getting close now.

File: src/psyclone/psyir/symbols/symboltable.py [hunk 7]
```diff
@@ -913,6 +931,20 @@ def tags_dict(self):
         '''
         return self._tags
 
+    def get_reverse_tags_dict(self):
+        '''
+        Constructs and returns a reverse of the map returned by
+        :py:method:`tags_dict`.
+
+        :returns: ordered dictionary of tags indexed by symbol.
+        :rtype: OrderedDict[:py:class:`psyclone.psyir.symbols.Symbol`] = str
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/symbols/symboltable.py:940]: I thought there might have been - I just copied it from somewhere else in the code :-) Either way, it seems like a sensible way to go.

[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/symbols/symboltable.py:940]: Updated as you suggest.

File: src/psyclone/psyir/symbols/symboltable.py [hunk 8]
```diff
@@ -913,6 +931,20 @@ def tags_dict(self):
...
+
+        '''
+        tags_dict_reversed = OrderedDict()
+        for tag, sym in self._tags.items():
+            tags_dict_reversed[sym] = tag
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/symbols/symboltable.py:945]: Oh yes, that did sort of float across my mind while I was doing this and then I forgot about it. I've extended the code and will add a test.

[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/symbols/symboltable.py:945]: In fact, implementing support for this means reworking other bits of the API so that a list of tags can be specified. I'm therefore going to suggest leaving this for now and have created a new Issue for it - #1654.

File: src/psyclone/psyir/transformations/hoist_local_arrays_trans.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+            # We must allow for the situation where there's a clash with a
+            # symbol name already present at container scope. (The validate()
+            # method will already have checked for tag clashes.)
+            try:
+                container.symbol_table.add(sym, tag=tags_dict.get(sym))
+            except KeyError:
+                new_name = container.symbol_table.next_available_name(
+                    sym.name, other_table=node.symbol_table)
+                node.symbol_table.rename_symbol(sym, new_name)
+                container.symbol_table.add(sym, tag=tags_dict.get(sym))
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/hoist_local_arrays_trans.py:160]: Oops. Well spotted. Fixed and extended the test.

File: src/psyclone/psyir/transformations/hoist_local_arrays_trans.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+            # support DataSymbols.
+            # pylint: disable=protected-access
+            del node.symbol_table._symbols[sym.name]
+
+    @staticmethod
+    def _get_local_arrays(node):
+        '''
+        Identify all arrays that are local to the target routine, do not
+        represent its return value and do not explicitly use dynamic memory
+        allocation.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/hoist_local_arrays_trans.py:195]: It's tested with the `all(isinstance(dim, ArrayType.ArrayBounds) for dim in sym.shape)` - an allocatable array will have array dimensions of ArrayType.Extent.DEFERRED. I'll add a comment.

[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/hoist_local_arrays_trans.py:195]: I've also updated the test as suggested.

File: src/psyclone/psyir/transformations/hoist_local_arrays_trans.py (hunk 6 above)
Comment (arporter) [reply src/psyclone/psyir/transformations/hoist_local_arrays_trans.py:283]: You and sympy are both right - that code was incorrect. Anyway, I like your suggestion of not failing so have moved this functionality to `_get_local_arrays()` and repurposed the existing test.

**Comment by arporter** [comment]
Thanks for spotting those problems Sergi. This is ready for another look if the CI is green.

**Comment by arporter** [comment]
Update: I've just re-tested with NEMOVAR (on CPU) and it still compiles and validates.

**Review by sergisiso — APPROVED** [review]
@arporter All comments have been addressed. Compilation tests, formatting and pylint and codecoverage all pass. The new transformation is in the user guide. Also it works and improves NEMO performance with the new changes. This is ready to merge.
